### ISSUE TITLE
extension member performance is suboptimal

### ISSUE THREAD

**Issue by jmesserly** [dart-lang/sdk#30531]
I was able to get significant improvement replacing `[dartx.add]` with a `__dart_add` ... 60% speedup on a microbenchmark that had a few symbol calls. It wasn't even designed to stress that.

I'm not sure if the problem is the Symbol, or that `dartx` is a big object and it requires extra work to fetch the symbol from it.

Switching to a non-symbol would risk affecting JS code, but it might be worth it for a large across the board improvement?

I have a CL ready that switches over to string property names.

**Comment by jmesserly** [comment]
Thoughts @vsmenon @leafpetersen on if the string approach is worth doing?

I could try to gather data on whether we can make symbols fast enough by skipping the `dartx` lookup.

Another improvement we can make, related to https://github.com/dart-lang/sdk/issues/30510: instead of tracking which types are extensions, we can track which member names are. So member names like `length` would always be treated as extensions. That would save load time, and simplify the compiler. Right now we set up the class to make both names available.

Note, this also suggests that we may want to investigate a non-symbol approach for private fields/members.

**Comment by jmesserly** [comment]
skipping `dartx.` was a 25% improvement, vs 60% for using string names.

**Comment by leafpetersen** [comment]
Ouch.  If symbols are slow, that's sadness for us.  We can probably mostly deal with the `dartx` issue by binding per module temps, right?  So each module would start with something like:

```js
const $bind = dartx.bind;
....
```

but the symbol vs string is potentially very large.  If you have a CL, I can try it out on some apps and tests.

Maybe we should check in with the v8 folks about symbol performance as well.

**Comment by leafpetersen** [comment]
Interestingly, these tests don't show a huge difference:

https://jsperf.com/symbol-vs-string-property-lookup-speed

Is there something about our pattern that is different?

**Comment by jmesserly** [comment]
yeah I implemented the `const $add = dartx.add;` pattern, that was the 25% speedup (vs 60%  for strings).

I don't think it's necessarily the symbol itself, it may be having to jump up several levels of the closure scope chain to find it (even though it's marked `const`). But that is just speculation on my part. It could also be that the benchmark is calling a tiny method and it ends up getting inlined, removing any costs associated with symbol lookup.

**Comment by jmesserly** [comment]
Symbol version is here: https://codereview.chromium.org/3002393002/

**Comment by mraleph** [comment]
Have you been able to figure out why `Symbol` based version is not as fast as string based one? (or why microbenchmark does not reflect this).

I am curious to take a look if you could send me the benchmark you are looking at.

**Comment by leafpetersen** [comment]
@jmesserly Do you have a benchmark you can send @mraleph or should I put something together?  I'd love to understand the underlying issue better - it might be a good thing to bring up with the V8 team.

**Comment by jmesserly** [comment]
RE: benchmark, what kind of benchmark are we talking about? What I've got is a Dart benchmark running against the Dart SDK, which is huge. And it relies on a module system of some sort, so I ran it in node.

Perhaps I can stitch together those JS files by hand, so it works in d8. The Dart benchmark wasn't designed to stress symbol-vs-string performance either, it was just something I tried out and was surprised by the large speed improvement.
